worker.go

```diff
@@ -3,373 +3,204 @@ package main
 import (
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
-	"log"
 	"os"
-	"os/exec"
 	"path/filepath"
-	"strconv"
 	"strings"
-	"sync"
 	"time"
-
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/awserr"
-	"github.com/aws/aws-sdk-go/service/s3"
 )
 
+type Job struct {
+	Root string
+	Dir  string
+}
+
 type Worker struct {
-	ID        int
-	Skipped   int64
-	wg        *sync.WaitGroup
-	jobQueue  <-chan Job
-	broadcast <-chan string
+	Uploader             FileUploader
+	DeleteFilesOnSuccess bool
+	Jobs                 <-chan Job
+	Results              chan<- string
+	Stop                 <-chan struct{}
 }
 
 type MetaData struct {
-	Name         string                 `json:"name"` // always set as "upload"
-	EpochNano    int64                  `json:"ts,omitempty"`
-	EpochNanoOld *int64                 `json:"timestamp,omitempty"` // only for reading (deprecated soon), keep it backwards compatible
-	Shasum       *string                `json:"shasum,omitempty"`
-	Labels       map[string]interface{} `json:"labels,omitempty"` // only read (will write to meta)
-	Meta         map[string]interface{} `json:"meta"`
-	Value        string                 `json:"val"` // only used to output URL, input is assumed to be empty
+	Name      string
+	Timestamp time.Time
+	Shasum    *string
+	Meta      map[string]string
+	Value     string
 }
 
-// ErrorStruct _
-type ErrorStruct struct {
-	Error string `json:"error,omitempty"`
+type UploadInfo struct {
+	NodeID    string
+	Namespace string
+	Name      string
+	Version   string
 }
 
-var sage_storage_api = "http://host.docker.internal:8080"
-var sage_storage_token = "user:test"
-var sage_storage_username = "test"
-
 func (worker *Worker) Run() {
-	defer worker.wg.Done()
-	fmt.Printf("Worker %d starting\n", worker.ID)
+	for job := range worker.Jobs {
```

Comment:
do u want a `defer` for `worker.Stop` here?

Reply:
Not in this case. All the components shared a common "stop" channel which is used to signal that they should stop. Basically, this allows us to stop everything by closing a single channel.

I did add a few comments to the Worker type to make this more clear, though.